### PR TITLE
Use FFTW on Windows

### PR DESCRIPTION
**Pull request by jakirkham**
Now that there are CMake-based builds of FFTW for Windows in conda-forge, try using them in the VIGRA build.

### CODE REVIEW

**Comment by jakirkham** [comment]
Any idea why we are seeing a missing symbol error with conda-forge's Windows FFTW and symbol `_fftwf_plan_many_dft`, @jschueller?

ref: https://ci.appveyor.com/project/conda-forge/vigra-feedstock/build/1.0.8/job/pmko4kx555n60el2#L2937

**Comment by jschueller** [comment]
seems implemented in "api/plan-many-dft.c" which is included.
Maybe vigra doesnt pas the FFTW_DLL define ? That's required when using the dll.
Maybe you could try set CMAKE_VERBOSE_MAKEFILE ?

**Comment by jschueller** [comment]
Also there's no need to bump as the unix package doesnt change.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ source:
 
 build:
   skip: true  # [win and py<35]
-  number: 1
+  number: 2
```
Comment (jschueller) [inline recipe/meta.yaml]: No need to bump as the unix package doesnt change.

Comment (jakirkham) [inline recipe/meta.yaml]: It's changing the Windows package though. So I have to bump it to get a new package for Windows.

**Comment by jakirkham** [comment]
It's changing the Windows package though. So I have to bump it to get a new package for Windows.

**Comment by jschueller** [comment]
oh, I thought there was none, sorry

**Comment by jakirkham** [comment]
Makes sense. No problem. Yeah just trying to add more functionality here for Windows users.

**Comment by jakirkham** [comment]
> Maybe vigra doesnt pass the FFTW_DLL define ?

So I defined `FFTW_DLL` in the C++ flags, but it looks like I'm getting a similar (though not the same) unresolved external symbol error. This time it is with `__imp__fftwf_plan_many_dft`. This [SO question]( https://stackoverflow.com/q/14776337 ) looks similar, but the [answer]( https://stackoverflow.com/a/14911790 ) looks pretty hacky to me. Any other ideas on how to proceed?

**Comment by jschueller** [comment]
It seems it wants to use fftw3f but actually links to fftw3, maybe set FFTW3_LIBRARY=PREFIX/Library/lib/fftw3f.lib

**Comment by jschueller** [comment]
Having the cmake verbose option set would help.

**Comment by jakirkham** [comment]
FWICT it finds [both `fftw3.lib` and `fftw3f.lib`]( https://ci.appveyor.com/project/conda-forge/vigra-feedstock/build/1.0.9/job/arp8e7tkeive37mt#L2511 ). Also it should use [both]( https://github.com/ukoethe/vigra/blob/Version-1-11-1/vigranumpy/src/fourier/CMakeLists.txt#L7-L8 ) to link this library.

Have enabled verbosity in a [build on my fork]( https://ci.appveyor.com/project/jakirkham/vigra-feedstock/build/1.0.42/job/u9pblrxqh2l6koq4#L3345 ). Seems to be linking to [both FFTW libraries]( https://ci.appveyor.com/project/jakirkham/vigra-feedstock/build/1.0.42/job/u9pblrxqh2l6koq4#L3361 ). Is there something about this linking that seems amiss?

**Comment by jschueller** [comment]
Could we use nmake VERBOSE=1 to see if cxx flags are actually passed ?

**Comment by jakirkham** [comment]
Am doing that [here]( https://github.com/jakirkham/vigra-feedstock/commit/4371ede29c60f0ede262b528844c30b0b8f32935#diff-d04c86b6bb20341f5f7c53165501a393R18 ). Did the [build]( https://ci.appveyor.com/project/jakirkham/vigra-feedstock/build/1.0.42/job/u9pblrxqh2l6koq4#L3345 ) on my fork as I didn't want to use up CI time while debugging this if I could help it.

**Comment by jschueller** [comment]
I dont see any flags passed to cl.exe

**Comment by jschueller** [comment]
I think it doesnt make sense to link both fftw3f and fftw3:
http://www.fftw.org/fftw2_doc/fftw_6.html:
"The names of FFTW functions, data types, and constants remain unchanged! You still call, for instance, fftw_one and not dfftw_one. Only the names of header files and libraries are modified. One consequence of this is that you cannot use both the single- and double-precision FFTW libraries in the same program, simultaneously, as the function names would conflict. "

Maybe the compiler is confused about with library to link.
What about setting FFTW3F_LIBRARY to fftw3.lib ?

**Comment by jakirkham** [comment]
Not sure what to say. Definitely passing [`VERBOSE=1`]( https://ci.appveyor.com/project/jakirkham/vigra-feedstock/build/1.0.42/job/u9pblrxqh2l6koq4#L2571 ).

**Comment by jakirkham** [comment]
> I think it doesnt make sense to link both fftw3f and fftw3

I'm not sure this is true. We link to both on Mac and Linux without issues. Though there could be something odd about Windows that is popping up here that I don't fully understand.

@ukoethe, am trying to build VIGRA with FFTW support on Windows. We have built FFTW using CMake so as to use the same runtime. However it seems we are running into linking issues. Have you seen an error like this before? Do you have any advise on how to proceed?

**Comment by jschueller** [comment]
Maybe unix compiler allow that as they allow undefined symbols in shared libs.
If the symbol names are the same, how does it know which one to use ? the first in ld path ?

**Comment by jschueller** [comment]
turns out symbol names are different after all:
http://www.fftw.org/fftw3_doc/Precision.html

**Comment by ukoethe** [comment]
The fourier module in vigranumpy must link against both `libfftw-3.3.lib` and `libfftwf-3.3.lib`. Normally, cmake is supposed to take care of this, and the cmake log says
```
-- Found FFTW3: C:/bld/vigra_1499697137864/_b_env/Library/lib/fftw3.lib  
-- Found FFTW3F: C:/bld/vigra_1499697137864/_b_env/Library/lib/fftw3f.lib
```
VIGRA expects these libraries to be provided by the variables `${FFTW3_LIBRARIES}` and `${FFTW3F_LIBRARIES}`. To debug, you should first check the contents of these variables when `vigranumpy/src/fourier/CMakeList.txt` is parsed.

**Comment by ukoethe** [comment]
If the linker picks up the right libraries, the next thing to check are the exported symbols. For example, if some `dllexport` was missing during the FFTW build, symbols may be absent from the libraries.

**Comment by jakirkham** [comment]
Toggling to restart CIs now that a new build of FFTW is out.

**Comment by jakirkham** [comment]
Thanks for following up @ukoethe. We were able to verify from a debug build in my branch that it does indeed link against both libraries. @jschueller noted some issues in the FFTW Windows build that might cause missing symbols. FFTW was rebuilt with that fix earlier today. Based on the preliminary results of the tests here, it seems like that will fix the issue. Will keep an eye on the builds here to see how it progresses, but expect the issue has been resolved.

**Comment by jakirkham** [comment]
Thanks for your help @jschueller and @ukoethe. Looks like this now passes. So have merged it so that we can get FFTW-based VIGRA builds out on Windows.

**Comment by jakirkham** [comment]
All builds are now up. Feedback welcome. 😄

**Comment by ukoethe** [comment]
It downloaded `https://conda.anaconda.org/conda-forge/win-64/vigra-1.11.1-np113py36_2.tar.bz2` (md5: 53e169b702df7ade11847738a4c229a2) and successfully executed the vigranumpy test suite on my Windows box (except for `test_arraytypes.py`, because I was too lazy to compile `vigranumpytest.pyd`, which is needed for this test).

**Comment by jschueller** [comment]
do these test stress the fftw part ?

**Comment by ukoethe** [comment]
No, but I checked manually that `fourierTransform()` and `fourierTransformInverse()` work.

**Comment by jschueller** [comment]
thanks. I guess this confirms our fftw build is ok.

**Comment by ukoethe** [comment]
Actually, vigra only calls the `float32` fftw version ATM. Theoretically, there could still be a problem with the `float64` variant, but it's not very likely.

**Comment by jakirkham** [comment]
FWIW we also run all the Python tests before releasing. So if there were Python tests covering those functions, we ran them.

Initially wanted to get all C++ tests too, but that took too much time on the CIs the last time I tried.

**Pull request merged**